Snowflake.Data/Core/Session/SessionPool.cs

```diff
@@ -187,15 +201,19 @@ internal bool AddSession(SFSession session)
                 return false;
             long timeNow = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
             if (session.IsNotOpen() || session.IsExpired(_timeout, timeNow))
+            {
+                lock (s_sessionPoolLock)
+                {
+                    _busySessionsCounter.Decrease();
+                }
                 return false;
+            }
 
             lock (s_sessionPoolLock)
             {
-                if (_idleSessions.Count >= _maxPoolSize)
-                {
-                    CleanExpiredSessions();
-                }
-                if (_idleSessions.Count >= _maxPoolSize)
+                _busySessionsCounter.Decrease();
+                CleanExpiredSessions();
+                if (GetCurrentPoolSize() >= _maxPoolSize)
```

Comment:
for me this method could be renamed to ClearIdleSessions() to avoid confusion.
So the question is if session is opened and is executing a query and we clear its pool, what should happen to the busy session.
For now I would do nothing about it. If in future we make the session aware of if it comes from a pool or not then we can think about a different behaviour.